rq/worker.py

```diff
@@ -1378,3 +1388,199 @@ class RandomWorker(Worker):
 
     def reorder_queues(self, reference_queue):
         shuffle(self._ordered_queues)
+
+
+class ThreadPoolWorker(Worker):
+    death_penalty_class = TimerDeathPenalty
+
+    def __init__(self, *args, **kwargs):
+        self.threadpool_size = kwargs.pop('pool_size', 24)
```

Comment:
I was just thinking a bit more about this. I think we can bring in attributes like `threadpool_size` and `_idle_threads` used here into the main `Worker` class, so they can be used by other types of workers implementing their own concurrency methods.

The default worker class will simply have `1` for `threadpool_size`, `_idle_threads` can also be incremented and deducted as work is being done.

I'd like to find a better and more generic name for `threadpool_size` though.

This way your work can go into RQ's master branch bit by bit. And if this works and `WorkerPool` is no longer needed, I'm also quite ok with it.

Reply:
I guess there are a couple of ways to go about this (thinking of multiple workers and future maintainability): (a) a "super" worker class, that does a lot and each worker implements very specific parts of different working models or (b) the minimum-denominator approach, where the base worker implements methods that all workers (regardless of working model) will use.

The current codebase uses strategy (a) - there are quite a few methods on the `Worker` that are not used by the `SimpleWorker` for example.

I would tend to favor strategy (b) - I think smaller classes are easier to reason about, simpler to maintain, and we "always" know that whatever is in the worker class is useful for any other worker, which is one of the main reasons why I think of the worker that forks a `ForkWorker` (we talked a bit about this on the V2 discussion).

No right or wrong here, so a choice of different approaches, both work, happy to adjust accordingly.
--

About the `WorkerPool`, I think that its complementary to a `ThreadWorker`. I think the main decision here is whether this will wrap the whole worker, or maybe just the `pop & work` part of things.